Marlin/src/core/ctstring.h

```diff
@@ -0,0 +1,98 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ */
+#pragma once
+
+#include <stddef.h>
+
+// This file contains few compile time string helpers functions to help reduce the binary
+// and runtime resources when working with string
+namespace CompileTimeString {
+  // Hash methods
+  //////////////////////////////////////////////////////////
+
+  // When to use those ?
+  //
+  // Typically, when doing a long chain of
+  // if (strcmp(txt, "ABC") == 0) {}
+  // else if (strcmp(txt, "CDE") == 0) {}
+  // etc...
+  // You're:
+  //   1. Storing the "ABC" / "CDE" string in the firmware
+  //   2. Using a O(N*M) algorithm (you are parsing the txt string M times with size N) while a O(1) is only required
+  //   3. Preventing the compiler to use a jump table (so larger code size is generated for all cases)
+
+  // So instead, do this:
+  // switch (constHash(txt)) {
+  //  case "ABC"_hash: {}
+  //  case "CDE"_hash: {}
+  // In that case, you're only paying for a O(N) algorithm here (computing the hash of txt)
+  // and then a O(1) for the switch/case jump table.
+  // This works because the case's hash are computed at compile time and only the hash is stored in the binary
+
+  // When not to use those ?
+  //
+  // If you have single:
+  // if (strcmp(A, B) == 0)
+  // Then adding the hash computation into the mix will actually slow things down, but you'll still gain
+  // the string storage size in flash. So it's worth it only if the string is larger than a machine word here (likely 4 bytes)
+
+  // Compute the hash of a string at compile time
+  size_t constexpr constHash(const char * const input) {
+    return *input ? static_cast<size_t>(*input) + 257 * constHash(input + 1) : 5381;
+  }
+
+  // String searching
+  //////////////////////////////////////////////////////////
+
+  // When to use those ?
+  //
+  // When you need to extract a small portion of a string that's known at compile time (can be automatically generated)
+  // Instead of doing runtime string operations, the compiler can deal with minimal string parsing at compile time
+  // Simple compile-time parser to find the position of the end of a string
+  constexpr const char* findStringEnd(const char * const str) {
+    return *str ? findStringEnd(str + 1) : str;
+  }
+
+  // Check whether a string contains a slash
+  constexpr bool containsSlash(const char * const str) {
+    return *str == '/' ? true : (*str ? containsSlash(str + 1) : false);
+  }
+  // Find the last position of the slash
+  constexpr const char* findLastSlashPos(const char * const str) {
+    return *str == '/' ? (str + 1) : findLastSlashPos(str - 1);
```

Comment:
This would be the more complete version. Should work, yes?
```cpp
    return *str == '\0' ? nullptr : (*str == '/' ? (str + 1) : findLastSlashPos(str - 1));
```

Reply:
No. The method search backward (from the end position). If you add this check it, it exits immediately.